jormungandr/src/blockchain/checkpoints.rs

```diff
@@ -0,0 +1,70 @@
+use crate::{blockcfg::HeaderHash, blockchain::Ref};
+use std::sync::Arc;
+
+/// list of pre-computed checkpoints from a given [`Ref`].
+///
+/// [`Ref`]: ./struct.Ref.html
+#[derive(Clone, Debug, PartialEq, Eq)]
+pub struct Checkpoints(Vec<HeaderHash>);
+
+impl Checkpoints {
+    /// create a new list of checkpoints from the given starting point (tip).
+    ///
+    /// The function make use of the `Ref` object pointing to previous epoch in order
+    /// to populate _interestingly_ spaced checkpoints.
+    ///
+    /// For now the algorithm provide the current block, the parent, the last block of the
+    /// previous epoch, the last block of the epoch before that... until the block0.
+    pub fn new_from(from: Arc<Ref>) -> Self {
+        let mut checkpoints = vec![from.hash(), from.block_parent_hash().clone()];
+
+        let mut ignore_prev = 0;
+        let mut current_ref = from;
+        while let Some(prev_epoch) = current_ref.last_ref_previous_epoch() {
+            current_ref = Arc::clone(prev_epoch);
+
+            for _ in 0..ignore_prev {
```

Comment:
This walks the epochs linearly, but skips incremental numbers of them. As the blockchain grows, the time to compute this every time will grow as well.

Reply:
but it does not matter because the epoch grow slowly. 1 every 3 or 4 days. This will work for now and we can find more precise construction later on

Reply:
Should we have some sort of O(1)-ish random access by epoch number which can be used here?

Reply:
we have time for that really. Right now after 2 years of blockchain, there is 143 epochs : https://cardanoexplorer.com

Reply:
It's cheaper that way. Maintaining an indexed accessor to epoch or checkpoints would make things more difficult for very little gain compared to the current approach.

With the current approach from the multiverse and the HAMT we have both rollbacks and branch for free. And again, unless in a testing mode, you are not going to create that many epochs per year. As counter intuitive as it feels the linked list is way faster here (and more certain).